pkg/cloud/gcp/gcs_kms.go

```diff
@@ -0,0 +1,196 @@
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"hash/crc32"
+	"net/url"
+	"strings"
+
+	kms "cloud.google.com/go/kms/apiv1"
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/errors"
+	"google.golang.org/api/option"
+	kmspb "google.golang.org/genproto/googleapis/cloud/kms/v1"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+const gcsScheme = "gs"
+
+type gcsKMS struct {
+	kms                 *kms.KeyManagementClient
+	customerMasterKeyID string
+}
+
+var _ cloud.KMS = &gcsKMS{}
+
+func init() {
+	cloud.RegisterKMSFromURIFactory(MakeGCSKMS, gcsScheme)
+}
+
+type kmsURIParams struct {
+	credentials string
+	auth        string
+}
+
+func resolveKMSURIParams(kmsURI url.URL) kmsURIParams {
+	params := kmsURIParams{
+		credentials: kmsURI.Query().Get(CredentialsParam),
+		auth:        kmsURI.Query().Get(cloud.AuthParam),
+	}
+
+	return params
+}
+
+// MakeGCSKMS is the factory method which returns a configured, ready-to-use
+// GCS KMS object.
+func MakeGCSKMS(uri string, env cloud.KMSEnv) (cloud.KMS, error) {
+	if env.KMSConfig().DisableOutbound {
+		return nil, errors.New("external IO must be enabled to use GCS KMS")
+	}
+	kmsURI, err := url.ParseRequestURI(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract the URI parameters required to setup the GCS KMS session.
+	kmsURIParams := resolveKMSURIParams(*kmsURI)
+
+	// Client options to authenticate and start a GCS KMS session.
+	// Currently only accepting json of service account.
+	var credentialsOpt []option.ClientOption
+
+	switch kmsURIParams.auth {
+	case "", cloud.AuthParamSpecified:
+		if kmsURIParams.credentials == "" {
+			return nil, errors.Errorf(
+				"%s is set to '%s', but %s is not set",
+				cloud.AuthParam,
+				cloud.AuthParamSpecified,
+				CredentialsParam,
+			)
+		}
+
+		// Credentials are passed in base64 encoded, so decode the credentials first.
+		credentialsJSON, err := base64.StdEncoding.DecodeString(kmsURIParams.credentials)
+		if err != nil {
+			return nil, err
+		}
+
+		credentialsOpt = append(credentialsOpt, option.WithCredentialsJSON(credentialsJSON))
+	case cloud.AuthParamImplicit:
+		if env.KMSConfig().DisableImplicitCredentials {
+			return nil, errors.New(
+				"implicit credentials disallowed for gcs due to --external-io-implicit-credentials flag")
+		}
+		// If implicit credentials used, no client options needed.
+	default:
+		return nil, errors.Errorf("unsupported value %s for %s", kmsURIParams.auth, cloud.AuthParam)
+	}
+
+	ctx := context.Background()
+
+	kmc, err := kms.NewKeyManagementClient(ctx, credentialsOpt...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Check the given cmk to see if key version is given.
```

Comment:
For my edification, does this account for a different kind of key we expect users to pass in. If you think it is valuable for future readers, can you flesh out this comment to include a link to any user facing GCS KMS docs?

Reply:
Yeah it was a concern brought up by CMEK team. Depending on where you copy the key from on gcp, it will either give you the key with the version and without. It throws an error if you give it with the version so that's why I just strip it if it's present. I will definitely flesh out the comment though